feature(sct-agent): sct-agent MVP integration

dimakr opened the pull request:
Integrate sct-agent MVP into SCT, as an alternative to using SSH for command execution on DB nodes.

Refs: https://github.com/scylladb/qa-tasks/issues/1858

### Testing
<!-- Add links to Argus/Jenkins of test test done with this PR -->
<!-- This would help the reviewer to cross check what was tested, and and review the results as needed -->
- [x] tested locally with a new `test-cases/agent-test-aws.yaml` test configuration, that was added to serve as a sanity test for sct-agent - it is essentially the `pr-provision-test`-like configuration, but with sct-agent enabled, hence the commands on DB nodes are executed via the agent, not SSH.

The status of the running agent on a DB node (the agent is registered as systemd service):
```
scyllaadm@agent-test-dmitriy-db-node-cfa6fbca-2:~$ sudo systemctl status sct-agent
● sct-agent.service - SCT Agent - Command execution agent for Scylla Cluster Tests
     Loaded: loaded (/etc/systemd/system/sct-agent.service; enabled; preset: enabled)
     Active: active (running) since Mon 2025-10-20 21:46:39 UTC; 16min ago
   Main PID: 4314 (sct-agent)
      Tasks: 10 (limit: 18771)
     Memory: 238.7M (peak: 506.2M)
        CPU: 12.609s
     CGroup: /system.slice/sct-agent.service
             ├─4314 /usr/local/bin/sct-agent --config /etc/sct-agent/config.yaml
             ├─6360 dirmngr --homedir /tmp --daemon
             └─6363 gpg-agent --homedir /tmp --use-standard-socket --daemon

Oct 20 21:58:31 agent-test-dmitriy-db-node-cfa6fbca-2 sudo[7197]:     root : PWD=/ ; USER=root ; COMMAND=/usr/bin/grep ^SCYLLA_ARGS= /etc/default/scylla-server
Oct 20 21:58:31 agent-test-dmitriy-db-node-cfa6fbca-2 sudo[7197]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=0)
Oct 20 21:58:31 agent-test-dmitriy-db-node-cfa6fbca-2 sudo[7197]: pam_unix(sudo:session): session closed for user root
Oct 20 21:58:31 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: 2025/10/20 21:58:31 sct-agent[6ba45b84]: Command completed successfully: exit_code=0 duration=8ms cmd=/bin/bash -c sudo grep "^SCYLLA_ARGS=" /etc/default/scylla-server
Oct 20 21:58:31 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: [2025-10-20 21:58:31] GET /api/v1/commands/6ba45b84-7c56-4f02-a30b-54a83322693f 200 41.095µs 46.243.158.145
Oct 20 21:58:31 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: [2025-10-20 21:58:31] POST /api/v1/commands 200 585.193µs 46.243.158.145
Oct 20 21:58:31 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: 2025/10/20 21:58:31 sct-agent[1d412446]: PWD=/ ; COMMAND=/bin/bash -c cat /etc/scylla.d/cpuset.conf
Oct 20 21:58:31 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: 2025/10/20 21:58:31 sct-agent[1d412446]: Command completed successfully: exit_code=0 duration=2ms cmd=/bin/bash -c cat /etc/scylla.d/cpuset.conf
Oct 20 21:58:31 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: [2025-10-20 21:58:31] GET /api/v1/commands/1d412446-cb56-4d01-90f1-b69e7d9034dc 200 35.189µs 46.243.158.145
Oct 20 21:58:32 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: [2025-10-20 21:58:32] GET /health 200 147.001µs 46.243.158.145
``` 

Some logs of the agent service:
```
scyllaadm@agent-test-dmitriy-db-node-cfa6fbca-2:~$ journalctl -u sct-agent | grep -i 'command completed' -A5 -B5
...
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: 2025/10/20 21:52:23 sct-agent[ac5e6753]: PWD=/ ; COMMAND=/bin/bash -c sudo systemctl stop scylla-server.service
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sudo[6333]:     root : PWD=/ ; USER=root ; COMMAND=/usr/bin/systemctl stop scylla-server.service
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sudo[6333]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=0)
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sudo[6333]: pam_unix(sudo:session): session closed for user root
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: 2025/10/20 21:52:23 sct-agent[ac5e6753]: Command completed successfully: exit_code=0 duration=16ms cmd=/bin/bash -c sudo systemctl stop scylla-server.service
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: [2025-10-20 21:52:23] GET /api/v1/commands/ac5e6753-1430-4890-90fb-f2a87404cb90 200 34.252µs 46.243.158.145
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: [2025-10-20 21:52:23] POST /api/v1/commands 200 557.768µs 46.243.158.145
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: 2025/10/20 21:52:23 sct-agent[56f7425a]: PWD=/ ; COMMAND=/bin/bash -c sudo rm -rf /var/lib/scylla/data/*
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sudo[6335]:     root : PWD=/ ; USER=root ; COMMAND=/usr/bin/rm -rf /var/lib/scylla/data/*
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sudo[6335]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=0)
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sudo[6335]: pam_unix(sudo:session): session closed for user root
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: 2025/10/20 21:52:23 sct-agent[56f7425a]: Command completed successfully: exit_code=0 duration=9ms cmd=/bin/bash -c sudo rm -rf /var/lib/scylla/data/*
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: [2025-10-20 21:52:23] GET /api/v1/commands/56f7425a-6169-47c0-83e4-3c19b6ddeb5f 200 33.039µs 46.243.158.145
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: [2025-10-20 21:52:23] POST /api/v1/commands 200 547.074µs 46.243.158.145
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: 2025/10/20 21:52:23 sct-agent[a20e695c]: PWD=/ ; COMMAND=/bin/bash -c sudo find /var/lib/scylla/commitlog -type f -delete
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sudo[6337]:     root : PWD=/ ; USER=root ; COMMAND=/usr/bin/find /var/lib/scylla/commitlog -type f -delete
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sudo[6337]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=0)
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sudo[6337]: pam_unix(sudo:session): session closed for user root
Oct 20 21:52:23 agent-test-dmitriy-db-node-cfa6fbca-2 sct-agent[4314]: 2025/10/20 21:52:23 sct-agent[a20e695c]: Command completed successfully: exit_code=0 duration=10ms cmd=/bin/bash -c sudo find /var/lib/scylla/commitlog -type f -delete
```

### PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I added the relevant `backport` labels
- [x] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevant to this change (if needed)

fruch commented on sdcm/utils/agent_client.py:
```diff
@@ -0,0 +1,245 @@
+        self.hostname = hostname
+        self.port = port
+        self.api_key = api_key
+        self.base_url = f"http://{hostname}:{port}"
```
let consider using ssh tunnel for development until we have https

dimakr replied:
I tried some changes for this, but bumped into a problem - we need dedicated tunnels for each sct agent; and the number of ports on local machine, for tunnels to listen on, should be equal to number of DB nodes we run in a test.
So this would require adding some mechanism to dynamically allocate local ports, and then track them. 

Do we want to add this? Or maybe use the current MVP implementation (use public IP when running tests locally with agent enabled). And drop it when https is added?

@fruch @soyacz

fruch replied:
Let's start with using public for local development 

But we need a secure solution for the way forward, before we can apply the agent by default across the board. (Which we probably would do in phases)

dimakr replied:
created https://github.com/scylladb/qa-tasks/issues/1998 task for this

fruch commented:
@dimakr 

isn't that default should come from the configuration file ?
https://github.com/dimakr/sct-agent/blob/7096ef7827ae069863b133c93a515cb4b95260e5/internal/executor/executor.go#L38

dimakr commented on sdcm/remote/agent_cmd_runner.py:
```diff
@@ -0,0 +1,231 @@
+        print(result.stdout)  # "hello\\n"
+    """
+
+    def __init__(self, hostname: str, port: int = 8443, api_key: str = "default-api-key",
```
fixed

dimakr commented on sdcm/remote/agent_cmd_runner.py:
```diff
@@ -0,0 +1,231 @@
+        print(result.stdout)  # "hello\\n"
+    """
+
+    def __init__(self, hostname: str, port: int = 8443, api_key: str = "default-api-key",
```
removed

dimakr commented on sdcm/cluster.py:
```diff
@@ -468,7 +469,22 @@ def _terminate_node_in_argus(self):
     def _init_remoter(self, ssh_login_info):
-        self.remoter = RemoteCmdRunnerBase.create_remoter(**ssh_login_info)
+        agent_config = self.parent_cluster.params.get('agent')
+        if agent_config['enabled'] and 'db' in self.parent_cluster.node_type:
```
Sample agent config is defined in default config of SCT.

dimakr commented on sdcm/remote/agent_cmd_runner.py:
```diff
@@ -0,0 +1,231 @@
+        :return: indication if operation was successful
+        """
+        sudo_prefix = "sudo " if sudo else ""
+        result = self.run(f"{sudo_prefix}cat {src}", timeout=timeout, ignore_status=False, verbose=False)
```
the same as above comment

dimakr commented on sdcm/cluster.py:
```diff
@@ -3549,12 +3565,14 @@ def prepare_user_data(self, enable_auto_bootstrap=False):
                                                   test_config=self.test_config,
-                                                  install_docker=self.node_type == 'loader')
+                                                  install_docker=self.node_type == 'loader',
+                                                  install_agent=agent_config.get('enabled') and 'db' in self.node_type)
```
Actually `sdcm.sct_provision.region_definition_builder.DefinitionBuilder._get_user_data_objects` (which Azure inherits) already includes SctAgentUserDataObject in its list of user data objects. So no changes are needed. Azure backend already works with sct agent deployment.
There was still a small fix required for reuse cluster case, where we do not run cloud-init, but startup script instead. So a flag was added to indicate if we want to install sct-agent when running startup script.

fruch commented on sdcm/utils/sct_agent_installer.py:
```diff
@@ -0,0 +1,200 @@
+        StandardError=journal
+        SyslogIdentifier=sct-agent
+        LimitNOFILE=65536
+
```
Let's make sure we keep a tab of resources utilization of the agents, let's raise any issue for that.

After that we can think if pinning is needed or not

dimakr replied:
created https://github.com/scylladb/qa-tasks/issues/1999 task for this

soyacz commented:
In logs I see:
```
< t:2025-11-19 09:33:47,087 f:agent_cmd_runner.py l:118  c:AgentCmdRunner       p:DEBUG > Executing command on 10.4.10.117: cat /etc/os-release
< t:2025-11-19 09:33:47,592 f:base.py         l:249  c:AgentCmdRunner       p:DEBUG > <10.4.10.117>: PRETTY_NAME="Ubuntu 24.04.3 LTS"
< t:2025-11-19 09:33:47,592 f:base.py         l:249  c:AgentCmdRunner       p:DEBUG > <10.4.10.117>: .04"
< t:2025-11-19 09:33:47,592 f:base.py         l:249  c:AgentCmdRunner       p:DEBUG > <10.4.10.117>: S (Noble Numbat)"
< t:2025-11-19 09:33:47,592 f:base.py         l:249  c:AgentCmdRunner       p:DEBUG > <10.4.10.117>: bian
< t:2025-11-19 09:33:47,592 f:base.py         l:249  c:AgentCmdRunner       p:DEBUG > <10.4.10.117>: ://www.ubuntu.com/"
< t:2025-11-19 09:33:47,592 f:base.py         l:249  c:AgentCmdRunner       p:DEBUG > <10.4.10.117>: m/"
< t:2025-11-19 09:33:47,592 f:base.py         l:249  c:AgentCmdRunner       p:DEBUG > <10.4.10.117>: net/ubuntu/"
< t:2025-11-19 09:33:47,592 f:base.py         l:249  c:AgentCmdRunner       p:DEBUG > <10.4.10.117>: ms-and-policies/privacy-policy"
< t:2025-11-19 09:33:47,592 f:base.py         l:167  c:AgentCmdRunner       p:DEBUG > <10.4.10.117>: Command "cat /etc/os-release" finished with status 0
```
it's skewed this way in multiple places. What's the reason? Can you fix that?

dimakr commented:
Fixed log collection problem that was spotted in the [run](https://argus.scylladb.com/tests/scylla-cluster-tests/f7f2c084-dcc6-428a-9d87-2af81f269d76/logs).

After that there was a [failed test run](https://argus.scylladb.com/tests/scylla-cluster-tests/06be45d0-c77b-4741-bbac-ceb75dd0160c/details). Connection to agent on one of nodes was timing out during a Nemesis, the node was under resources pressure - the following messages from scylla appear at that time in logs:
```
Too long queue accumulated for sl:default (1126 tasks)
...
Too long queue accumulated (rate limiting dropped 83 similar messages)
``` 
and seems like agent was lacking resource because of that.

I added CPU pinning (which we planned to do anyway in the subsequent tasks) for agent service in `sdcm.utils.sct_agent_installer.get_agent_systemd_service`. The [latest run](https://argus.scylladb.com/tests/scylla-cluster-tests/ce789fa0-dade-454c-a428-5c3c55745fb3) was OK.

soyacz commented on defaults/test_default.yaml:
```diff
@@ -336,3 +336,11 @@ download_from_s3: []
+agent:
+  enabled: false
+  port: 16000
+  binary_url: ""
```
why not providing a default url? How one may know where to get it?

fruch commented on sdcm/cluster.py:
```diff
@@ -534,8 +542,61 @@ def _terminate_node_in_argus(self):
+        if agent_enabled and agent_api_key:
+            agent_hostname = (
+                self.public_ip_address or self.external_address
+                if self.test_config.IP_SSH_CONNECTIONS == "public"
```
we can take it out, and whom would need it for local testing, should be able to figure it out.
